### PR TITLE
Feature/user preferences module

### LINKED ISSUE #78: DB Model - User preferences
**Is your feature request related to a problem? Please describe.**
As a regular user of PCM, I want to save my preferences.

**Describe the solution you'd like**
There should be a DB model for User preferences (One to One relationship) with properties:
- [ ] `userId` (UUID)
- [ ] `practiceGoal` (Number 1-4)
- [ ] `why` (Text field of why they are here)

The `UserService` should be appended to include:
- [ ] An `updatePreferences` method to set the user's preferences.

**Additional context**
- This model will be used by the `updatePreferences` mutation

### LINKED ISSUE #79: API Mutation - User Preferences
**Is your feature request related to a problem? Please describe.**
As someone new to the platform, I want to tell the platform about myself (why I'm here, my skill level, etc)

**Describe the solution you'd like**
There should be an API graphql query `updatePreferences` that allow the user to configure:
- [ ] Their coding ability (1-10)
- [ ] Why they are here (text field)
- [ ] Practice cadence (1-4)

**Additional context**
- Eventually there may be extra preferences we add, however for now let' keep it to these 3.
- Relies on #78

### PR DESCRIPTION
**Pull request by quimgv**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

## Fixed issues (PRs without this will not be accepted)
- Fixes #78 
- Fixes #79 

## Changes
- Added userPreferences module (entity and service)
- Added updatePreferences resolver
- Added updatePreferences FieldResolver in UserResolver
- Added updatePreferences test
- Moved setup to TestClient

## PR Checklist
- [x] Does your PR contain relevant tests that pass?
- [x] Is your code linted? `yarn lint` or `yarn lint:fix`
- [x] Does your code build correctly in ALL projects? `yarn build`

### CODE REVIEW

File: packages/api/src/User/User.resolver.ts [hunk 1]
```diff
@@ -27,4 +32,20 @@ export class UserResolver {
   createUser(@Args({ name: 'user', type: () => UserInput }) user: UserInput) {
     return this.userService.create(user);
   }
+
+  @UseGuards(GQLAuthGuard)
+  @Mutation(() => UserPreferences)
+  updatePreferences(@CurrentUser() user: User, @Args({ name: 'input', type: () => UserPreferencesInput }) input: UserPreferencesInput) {
```
Comment (tristanMatthias) [inline packages/api/src/User/User.resolver.ts:38]: Can you split this into multiple lines please?

File: packages/api/src/User/index.ts [hunk 2]
```diff
@@ -4,11 +4,13 @@ import { TypeOrmModule } from '@nestjs/typeorm';
...
 import { UserService } from './User.service';
+import { UserPreferences } from '../UserPreferences/UserPreferences.entity';
+import { UserPreferencesService } from '../UserPreferences/UserPreferences.service';
 
 
 @Module({
-  imports: [TypeOrmModule.forFeature([UserWithPassword])],
-  providers: [UserResolver, UserService, UserWithPassword],
+  imports: [TypeOrmModule.forFeature([UserWithPassword, UserPreferences])],
+  providers: [UserResolver, UserService, UserWithPassword, UserPreferencesService],
```
Comment (tristanMatthias) [inline packages/api/src/User/index.ts:13]: Same with these; break into new lines

File: packages/api/src/UserPreferences/UserPreferences.service.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+@Injectable()
+export class UserPreferencesService {
+  constructor(
+    @InjectRepository(UserPreferences)
+    private readonly userPreferencesRepository: Repository<UserPreferences>
+  ) { }
+
+  async findByUser(userId: string): Promise<UserPreferences | undefined> {
+    const userPreference = await this.userPreferencesRepository.findOne({ where: { userId } });
+    return userPreference;
```
Comment (tristanMatthias) [inline packages/api/src/UserPreferences/UserPreferences.service.ts:16]: combine into one line, unless we're doing something here

File: packages/api/src/UserPreferences/UserPreferences.service.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  }
+
+  async update(userId: string, input: UserPreferencesInput): Promise<UserPreferences> {
+    const userPreferences = await this.userPreferencesRepository.findOne({ where: { userId } });
+    if (userPreferences) {
+      return this.userPreferencesRepository.save({ ...userPreferences, ...input });
+    }
+    return this.userPreferencesRepository.create({ userId, ...input }).save();
+  }
```
Comment (tristanMatthias) [inline packages/api/src/UserPreferences/UserPreferences.service.ts:25]: Is there an upsert in typeorm? If so let's us this, otherwise ignore this comment

File: packages/api/src/UserPreferences/index.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+import { Module } from '@nestjs/common';
+
+@Module({})
+export class UserPreferencesModule {}
```
Comment (tristanMatthias) [inline packages/api/src/UserPreferences/index.ts:4]: Is this actually used anywhere?

File: packages/api/tests/entities/Path.test.ts [hunk 6]
```diff
@@ -7,11 +7,6 @@ const pathInput: PathInput = {
   description: 'Description text'
 };
 
-const setup = async () => {
-  await TestClient.resetDatabase();
-  await TestClient.workflowSignup();
-};
-
```
Comment (tristanMatthias) [inline packages/api/tests/entities/Path.test.ts:14]: Undo this change, as different tests have different setups

File: packages/api/tests/utils/TestClient.ts [hunk 7]
```diff
@@ -57,6 +58,14 @@ export abstract class TestClient {
...
   }
 
+  /**
+   * Resets the database, creates a user and log the user in
+   */
+  static async setup() {
+    await this.resetDatabase();
+    await this.workflowSignup();
+  }
+
```
Comment (tristanMatthias) [inline packages/api/tests/utils/TestClient.ts:68]: Undo this, as per other comment

File: packages/api/src/UserPreferences/UserPreferences.entity.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    @Column()
+    @Field()
+    userId: string;
+
+    @Column()
+    @Field()
+    @IsInt()
+    @Min(1)
+    @Max(4)
+    practiceGoal: number;
```
Comment (Ellie-2020) [inline packages/api/src/UserPreferences/UserPreferences.entity.ts:22]: I do not understand why, but it seems still allows setting number other than 1 - 4 to database.

File: packages/api/src/UserPreferences/index.ts (hunk 5 above)
Comment (quimgv) [reply packages/api/src/UserPreferences/index.ts:4]: Not really, shall I remove it? I still don't completely get how imports/exports work some times... I can get the service from User entity without exporting anything. Maybe the relation between tables has something to do with that?

File: packages/api/src/UserPreferences/UserPreferences.service.ts (hunk 4 above)
Comment (quimgv) [reply packages/api/src/UserPreferences/UserPreferences.service.ts:25]: I didn't see any

File: packages/api/src/UserPreferences/UserPreferences.entity.ts (hunk 8 above)
Comment (tristanMatthias) [reply packages/api/src/UserPreferences/UserPreferences.entity.ts:22]: Hmm, interesting. This looks correct according to the offical docs. How are you testing it?

Comment (tristanMatthias) [reply packages/api/src/UserPreferences/UserPreferences.entity.ts:22]: @quimgv we should add a test for this

Comment (Ellie-2020) [reply packages/api/src/UserPreferences/UserPreferences.entity.ts:22]: On localhost playground, I did mutation:
`mutation {
  updatePreferences(input: {  
    practiceGoal: 7
    why: "secret 4"
    codingAbility: 20
})
  {
    id
    userId
    practiceGoal
    why
    codingAbility
  }
}`

I get successful result, and check the database, the value is set exactly as what is in the request, 7 and 20. 
`{
  "data": {
    "updatePreferences": {
      "id": "8fad1616-d9a2-4681-945c-6c2a21f11065",
      "userId": "421dd1a7-d5d1-4a36-9ef3-14e4a91cb920",
      "practiceGoal": 7,
      "why": "secret 4",
      "codingAbility": 20
    }
  }
}`

File: packages/api/src/UserPreferences/index.ts (hunk 5 above)
Comment (tristanMatthias) [reply packages/api/src/UserPreferences/index.ts:4]: If it's not used it can be deleted. The other user preference stuff is consumed directly by the database module. It appears that modules are just ways of grouping things in nest.

File: packages/api/src/UserPreferences/UserPreferences.entity.ts (hunk 8 above)
Comment (tristanMatthias) [reply packages/api/src/UserPreferences/UserPreferences.entity.ts:22]: OK great call. Thanks @Ellie-2020. We will definitely need a test for this then. Not sure why that's happening. Have a further dig, and if you guys can't find it I'll have a look later on

**Comment by quimgv** [comment]
@tristanMatthias It's only left to do the the validation issue.

@Ellie-2020 Good catch!

**Comment by quimgv** [comment]
@tristanMatthias FYI @Ellie-2020 has #86 done in local but relies on this and she needs to wait for this to be merged before creating the PR.

**Comment by tristanMatthias** [comment]
@quimgv , @Ellie-2020 I fixed this up. Read the last commit for details. If someone can add some tests to check this, that would be great, then I'll merge in

File: packages/api/src/UserPreferences/UserPreferences.entity.ts (hunk 8 above)
Comment (Ellie-2020) [reply packages/api/src/UserPreferences/UserPreferences.entity.ts:24]: I just pull the latest version of the branch and tested. Now it works. I did 3 test using mutation updateUserPreferences:
1. all the 3 parameters with valid value, it execute successful.
2. change practiceGoal value as 8 which is invalid. It failed as expected.

Comment (Ellie-2020) [reply packages/api/src/UserPreferences/UserPreferences.entity.ts:24]: 3. change  codingAbility value as 90 which is invalid. it failed as expected. 

So it is solved.

**Comment by Ellie-2020** [comment]
> @quimgv , @Ellie-2020 I fixed this up. Read the last commit for details. If someone can add some tests to check this, that would be great, then I'll merge in

Oops, I do not know how it happened, the comments was added to wrong place just now... I did test, and it now solved.

**Comment by tristanMatthias** [comment]
Ok, did you push those up? I can't see them on github
